### PR TITLE
SIRI-88: Implementation for CodeListEntries Import Job

### PR DESCRIPTION
**Pull request by MGrammet**

### CODE REVIEW

File: src/main/java/sirius/biz/jobs/params/CodeListParameter.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/*
```
Comment (andyHa) [inline src/main/java/sirius/biz/jobs/params/CodeListParameter.java]: Couldn't we rely on a generic parameter here using CodeLists and the unique name of a code list instead of the id? This would spare us from creating separate implementations.

Comment (MGrammet) [reply src/main/java/sirius/biz/jobs/params/CodeListParameter.java]: CodeListParameter is now generic.

File: src/main/java/sirius/biz/codelists/CodeListEntryData.java [hunk 2]
```diff
@@ -43,12 +44,14 @@
     @Trim
     @Length(50)
     @Unique(within = "codeList")
+    @AutoImport
```
Comment (andyHa) [inline src/main/java/sirius/biz/codelists/CodeListEntryData.java]: This must not ever change.

File: src/main/java/sirius/biz/codelists/CodeLists.java [hunk 3]
```diff
@@ -245,20 +253,30 @@ private L findOrCreateCodelist(@Nonnull String codeList) {
...
-        if (cl == null) {
+        return Optional.ofNullable(cl);
+    }
+
+    private L findOrCreateCodelist(@Nonnull String codeList) {
+        Optional<L> optionalCodeList = findCodelist(codeList);
+
+        Tenant<?> currentTenant = getCurrentTenant(codeList).orElseThrow(this::warnAboutMissingTenant);
+
+        if (!optionalCodeList.isPresent()) {
```
Comment (andyHa) [inline src/main/java/sirius/biz/codelists/CodeLists.java]: invert if an return immediatelly

File: src/main/java/sirius/biz/codelists/jdbc/SQLCodeListEntry.java [hunk 4]
```diff
@@ -22,9 +24,14 @@
...
 public class SQLCodeListEntry extends SQLEntity implements CodeListEntry<Long, SQLCodeList> {
 
+    /**
+     * Contains the codeList to which this SQLCodeListEntry belongs.
+     */
+    public static final Mapping CODE_LIST = Mapping.named("codeList");
     private final SQLEntityRef<SQLCodeList> codeList =
             SQLEntityRef.on(SQLCodeList.class, SQLEntityRef.OnDelete.CASCADE);
 
+    @AutoImport
```
Comment (andyHa) [inline src/main/java/sirius/biz/codelists/jdbc/SQLCodeListEntry.java]: ??

File: src/main/java/sirius/biz/codelists/jdbc/SQLCodeListEntryImportHandler.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+     * @param clazz   the type of entities being handled
+     * @param context the import context to use
+     */
+    protected SQLCodeListEntryImportHandler(Class<?> clazz, ImporterContext context) {
+        super(clazz, context);
+    }
+
+    @Override
+    protected void collectFindQueries(BiConsumer<Predicate<SQLCodeListEntry>, Supplier<FindQuery<SQLCodeListEntry>>> queryConsumer) {
+        queryConsumer.accept(sqlCodeListEntry -> !sqlCodeListEntry.isNew(),
```
Comment (andyHa) [inline src/main/java/sirius/biz/codelists/jdbc/SQLCodeListEntryImportHandler.java]: is there any good reason to support IDs?

File: src/main/java/sirius/biz/codelists/mongo/MongoCodeListEntryImportHandler.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+    /**
+     * Tries to find an entity using the supplied <tt>data</tt>.
+     *
+     * @param data the data used to describe the entity to find
+     * @return a matching entity wrapped as optional or an empty optional if there is no matching entity
+     */
+    @Override
+    public Optional<MongoCodeListEntry> tryFind(Context data) {
+        if (data.containsKey(MongoCodeListEntry.ID.getName())) {
```
Comment (andyHa) [inline src/main/java/sirius/biz/codelists/mongo/MongoCodeListEntryImportHandler.java]: s.a.

File: src/main/java/sirius/biz/jobs/params/CodeListParameter.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+import java.util.Optional;
+
+/**
+ * Permits to select a {@link CodeList} as parameter.
+ *
+ * @param <I> the type of the ID used by subclasses.
+ * @param <V> the type of selectable by this parameter.
+ */
+public class CodeListParameter<I, V extends BaseEntity<I> & CodeList> extends Parameter<V, CodeListParameter<I, V>> {
```
Comment (andyHa) [inline src/main/java/sirius/biz/jobs/params/CodeListParameter.java]: no need for generics here(?)

File: src/main/java/sirius/biz/jobs/params/CodeListParameter.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+    public static final String CODE_LISTS_AUTOCOMPLETE = "/code-lists/autocomplete";
```
Comment (andyHa) [inline src/main/java/sirius/biz/jobs/params/CodeListParameter.java]: javadoc or private

Comment (andyHa) [reply src/main/java/sirius/biz/jobs/params/CodeListParameter.java]: can't we directly use this URI in the template?

File: src/main/java/sirius/biz/jobs/params/CodeListParameter.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+        return (Optional<V>) codeLists.findCodelist(input.getString());
+    }
+
+    /**
+     * Returns the autocompletion URL used to determine suggestions for inputs provided by the user.
+     *
+     * @return the autocomplete URL used to provide suggestions for user input
+     */
+    public String getAutocompleteUri() {
```
Comment (andyHa) [inline src/main/java/sirius/biz/jobs/params/CodeListParameter.java]: s.a.

File: src/main/java/sirius/biz/codelists/jdbc/SQLCodeListEntry.java [hunk 10]
```diff
@@ -22,9 +24,14 @@
 @TranslationSource(CodeListEntry.class)
 public class SQLCodeListEntry extends SQLEntity implements CodeListEntry<Long, SQLCodeList> {
 
+    /**
+     * Contains the codeList to which this SQLCodeListEntry belongs.
+     */
+    public static final Mapping CODE_LIST = Mapping.named("codeList");
```
Comment (andyHa) [inline src/main/java/sirius/biz/codelists/jdbc/SQLCodeListEntry.java]: redundant - defined in CodeListEntry

File: src/main/java/sirius/biz/jobs/params/CodeListParameter.java [hunk 11]
```diff
@@ -19,13 +18,8 @@
...
 /**
  * Permits to select a {@link CodeList} as parameter.
- *
- * @param <I> the type of the ID used by subclasses.
- * @param <V> the type of selectable by this parameter.
  */
-public class CodeListParameter<I, V extends BaseEntity<I> & CodeList> extends Parameter<V, CodeListParameter<I, V>> {
-
-    public static final String CODE_LISTS_AUTOCOMPLETE = "/code-lists/autocomplete";
+public class CodeListParameter extends Parameter {
```
Comment (andyHa) [inline src/main/java/sirius/biz/jobs/params/CodeListParameter.java]: Parameter<CodeList, CodeListParameter>

File: src/main/java/sirius/biz/codelists/mongo/MongoCodeListEntryImportJobFactory.java [hunk 12]
```diff
@@ -63,20 +62,29 @@ protected void collectParameters(Consumer<Parameter<?, ?>> parameterCollector) {
...
+        @SuppressWarnings("unchecked")
         private MongoCodeListEntryImportJob(ProcessContext process) {
             super(fileParameter,
                   ignoreEmptyParameter,
                   importModeParameter,
                   MongoCodeListEntry.class,
                   getDictionary(),
                   process);
-            codeList = process.require(codeListParameter);
+            codeList = (MongoCodeList) process.require(codeListParameter);
```
Comment (andyHa) [inline src/main/java/sirius/biz/codelists/mongo/MongoCodeListEntryImportJobFactory.java]: use interface instead of concrete class

File: src/main/java/sirius/biz/codelists/mongo/MongoCodeListEntryImportJobFactory.java [hunk 13]
```diff
@@ -63,20 +62,29 @@ protected void collectParameters(Consumer<Parameter<?, ?>> parameterCollector) {
          *
          * @param process the process context itself
          */
+        @SuppressWarnings("unchecked")
```
Comment (andyHa) [inline src/main/java/sirius/biz/codelists/mongo/MongoCodeListEntryImportJobFactory.java]: remove

File: src/main/java/sirius/biz/codelists/jdbc/SQLCodeListEntryImportJobFactory.java [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+
+    @Override
+    protected void computePresetFor(Object targetObject, Map<String, Object> preset) {
+        preset.put(codeListParameter.getName(), ((SQLCodeList) targetObject).getId());
+    }
+
+    @Override
+    protected void enhanceDictionary(ImportDictionary dictionary) {
+        super.enhanceDictionary(dictionary);
+        FieldDefinition code =
```
Comment (andyHa) [inline src/main/java/sirius/biz/codelists/jdbc/SQLCodeListEntryImportJobFactory.java]: why is this needed?
